flink/v1.12/flink/src/test/java/org/apache/iceberg/flink/source/TestProjectMetaColumn.java

```diff
@@ -0,0 +1,180 @@
+public class TestProjectMetaColumn {
+
+  @Rule
+  public final TemporaryFolder folder = new TemporaryFolder();
+  private final FileFormat format;
+
+  @Parameterized.Parameters(name = "fileFormat={0}")
+  public static Iterable<Object[]> parameters() {
+    return Lists.newArrayList(
+        new Object[] {FileFormat.PARQUET},
+        new Object[] {FileFormat.ORC},
+        new Object[] {FileFormat.AVRO}
+    );
+  }
+
+  public TestProjectMetaColumn(FileFormat format) {
+    this.format = format;
+  }
+
+  private void testSkipToRemoveMetaColumn(int formatVersion) throws IOException {
+    // Create the table with given format version.
+    String location = folder.getRoot().getAbsolutePath();
+    Table table = SimpleDataUtil.createTable(location,
+        ImmutableMap.of("format-version", String.valueOf(formatVersion)),
+        false);
+
+    List<RowData> rows = Lists.newArrayList(
+        SimpleDataUtil.createInsert(1, "AAA"),
+        SimpleDataUtil.createInsert(2, "BBB"),
+        SimpleDataUtil.createInsert(3, "CCC")
+    );
+    writeAndCommit(table, ImmutableList.of(), false, rows);
+
+    FlinkInputFormat input = FlinkSource
+        .forRowData()
+        .tableLoader(TableLoader.fromHadoopTable(location))
+        .buildFormat();
+
+    List<RowData> results = Lists.newArrayList();
+    TestHelpers.readRowData(input, rowData -> {
+      // If project to remove the meta columns, it will get a RowDataProjection.
+      Assert.assertTrue(rowData instanceof GenericRowData);
+      results.add(TestHelpers.copyRowData(rowData, SimpleDataUtil.ROW_TYPE));
+    });
+
+    // Assert the results.
+    TestHelpers.assertRows(rows, results, SimpleDataUtil.ROW_TYPE);
+  }
+
+  @Test
+  public void testV1SkipToRemoveMetaColumn() throws IOException {
+    testSkipToRemoveMetaColumn(1);
+  }
+
+  @Test
+  public void testV2SkipToRemoveMetaColumn() throws IOException {
+    testSkipToRemoveMetaColumn(2);
+  }
+
+  @Test
+  public void testV2RemoveMetaColumn() throws Exception {
+    // Create the v2 table.
+    String location = folder.getRoot().getAbsolutePath();
+    Table table = SimpleDataUtil.createTable(location, ImmutableMap.of("format-version", "2"), false);
+
+    List<RowData> rows = Lists.newArrayList(
+        SimpleDataUtil.createInsert(1, "AAA"),
+        SimpleDataUtil.createDelete(1, "AAA"),
+        SimpleDataUtil.createInsert(2, "AAA"),
+        SimpleDataUtil.createInsert(2, "BBB")
+    );
+    int eqFieldId = table.schema().findField("data").fieldId();
+    writeAndCommit(table, ImmutableList.of(eqFieldId), true, rows);
+
+    FlinkInputFormat input = FlinkSource
+        .forRowData()
+        .tableLoader(TableLoader.fromHadoopTable(location))
+        .buildFormat();
+
+    List<RowData> results = Lists.newArrayList();
+    TestHelpers.readRowData(input, rowData -> {
+      // If project to remove the meta columns, it will get a RowDataProjection.
+      Assert.assertTrue(rowData instanceof RowDataProjection);
```

Comment:
For v2 table with eq-deletes or pos-deletes to merge,  we will add `_pos`, `file_path`, `is_deleted` meta columns.  For flink engine,  we will need to project the row data to remove the extra meta columns for users.  The `RowDataProjection` was introduced since here https://github.com/apache/iceberg/pull/3240/files#diff-aed36b23cf7e6110f84bd3b34307240d2174b2be2edb987dfcd22ff15ab75f72R81-R83